tests/FSharp.Core.UnitTests/LibraryTestFx.fs

```diff
@@ -18,7 +18,7 @@ let CheckThrowsExn<'a when 'a :> exn> (f : unit -> unit) =
         let _ = f ()
         sprintf "Expected %O exception, got no exception" typeof<'a> |> Assert.Fail 
     with
-    | :? 'a -> ()
+    | :? 'a as e when e.GetType() = typeof<'a> -> ()
```

Comment:
TBH I do not think that it is wrong when an API promises is to throw exception A, and instead it returns B which inherits A.
This should be an implementation detail, and as such should probably not cause tests to fail.

Reply:
I assume this is what you used when figuring out which tests to change in the first place.
But now that it has happened, I think the whole when guard can be removed.

Reply:
That was my first thought also. But now I believe it rather applies to the production code (and hence the operator :? was likely designed to follow the inheritance chain), whereas in these test frameworks we should be more precise.

Getting back to the original trigger (the PR I was reviewing), we have the function taking an argument and we want to:
1) Throw ArgumentNullException for value X
2) Throw ArgumentOutOfBoundsException for value Y
3) Throw ArgumentException for value Z

I want my tests to be expressive about that with code like
```fsharp
CheckThrowsArgumentNullException (fun () -> (fun () -> func(X))
CheckThrowsArgumentOutOfBoundsException (fun () -> func(Y))
CheckThrowsArgumentException (fun () -> func(Z))
```
If this above does NOT actually check the exact exception, how to ensure the precise behavior? We'd either need to review code manually or force end tests to add a type check to every case like this.

To sum up, I think for the case of test frameworks, this precise check is beneficial.

Reply:
It can check the exception and allow either that OR allow any of its inherited types, because LSP should still hold here.

If it checks CheckThrowsArgumentOutOfBoundsException and something which inherits ArgumentOutOfBoundsException  -> that's OK.

Reply:
If this is kept and lib code is reimplemented, while API still holds - this is just sunk costs to update the tests, even when there is nothing wrong with such code.
The test should only fail if something goes wrong in the impl, not when it changes internal implementation detail.

Reply:
Took me a while to understand why you're mentioning LSP :D

No I understand what you mean, it's more important that we detect when `ArgumentException` is thrown instead of `ArgumentOutOfBoundsException` and not the opposite case - and the current (unmodified) impl does check this.

I just don't see anything bad in being more precise here as it will help avoid similar situations in the future and will force tests to read closer to the spec.

Reply:
Because there is nothing "precise" about it at all.
If API says it returns `A`, and instead it returns `B : A` - that is still very precise meeting of the spec, and this is the case across many platforms incl. dotnet.

The only thing insisting on exact type and disallowing subtypes can cause is additional effort for people who would implement an actual good change and the tests would get in the way.

And yes, it is this LSP which was meant by it: https://en.wikipedia.org/wiki/Liskov_substitution_principle :))

Reply:
So - not convinced, but 2:1 and the point was to fix the tests, so let's keep things as they are :)